### PR TITLE
Support to change the policy configmap for k8s-keystone-auth on the fly

### LINKED ISSUE #195: Improvements for k8s-keystone-auth
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind feature

**What happened**:

**What you expected to happen**:
In a perfect world, I would like to see:
1. k8s-keystone-auth can be run either standalone or as pod
2. k8s-keystone-auth can accept both policy file or configMap
3. Policies can be updated and effected on-the-fly
4. If k8s-keystone-auth is running as pod, it can work on worker node.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- openstack-cloud-controller-manager version:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:
  This is related to Magnum/k8s-keystone-auth integration.

### PR DESCRIPTION
**Pull request by lingxiankong**
**What this PR does / why we need it**:
Support to change the policy configmap for k8s-keystone-auth on the fly

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #195

**Special notes for your reviewer**:
- Watch the configmap change for authorization policy and apply to the
  service dynamically.
- Improve the shape of k8s-keystone-auth implementation

**Release note**:

```release-note
The authorization functionality of k8s-keystone-auth service can be running either by specifying a policy file or a configmap in kube-system namespace. If the configmap is provided, the authorization policy could be changed during the service running without service restart.
```

### CODE REVIEW

File: cmd/k8s-keystone-auth/main.go [hunk 1]
```diff
@@ -15,82 +15,33 @@ limitations under the License.
...
 
 func main() {
-	flag.CommandLine.Parse([]string{})
-	pflag.StringVar(&listenAddr, "listen", "0.0.0.0:8443", "<address>:<port> to listen on")
-	pflag.StringVar(&tlsCertFile, "tls-cert-file", os.Getenv("TLS_CERT_FILE"), "File containing the default x509 Certificate for HTTPS.")
-	pflag.StringVar(&tlsPrivateKey, "tls-private-key-file", os.Getenv("TLS_PRIVATE_KEY_FILE"), "File containing the default x509 private key matching --tls-cert-file.")
-	pflag.StringVar(&keystoneURL, "keystone-url", os.Getenv("OS_AUTH_URL"), "URL for the OpenStack Keystone API")
-	pflag.StringVar(&keystoneCaFile, "keystone-ca-file", os.Getenv("KEYSTONE_CA_FILE"), "File containing the certificate authority for Keystone Service.")
-	pflag.StringVar(&policyFile, "keystone-policy-file", os.Getenv("KEYSTONE_POLICY_FILE"), "File containing the policy, if provided, it takes precedence over the policy configmap.")
-	pflag.StringVar(&configMap, "policies-configmap-name", "", "ConfigMap in kube-system namespace containing the policy configuration, the ConfigMap must contain the key 'policies'")
```
Comment (zioproto) [inline cmd/k8s-keystone-auth/main.go]: Should this be policies or policy ?

File: pkg/identity/keystone/config.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+	if c.KeystoneURL == "" {
+		errorsFound = true
+		glog.Errorf("please specify --keystone-url or set the OS_AUTH_URL environment variable.")
+	}
+	if c.CertFile == "" || c.KeyFile == "" {
+		errorsFound = true
+		glog.Errorf("Please specify --tls-cert-file and --tls-private-key-file arguments.")
+	}
+	if c.PolicyFile == "" && c.PolicyConfigMapName == "" {
```
Comment (zioproto) [inline pkg/identity/keystone/config.go]: policy or policies ?

File: pkg/identity/keystone/config.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+// AddFlags adds flags for a specific AutoScaler to the specified FlagSet
+func (c *Config) AddFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&c.Address, "listen", c.Address, "<address>:<port> to listen on")
+	fs.StringVar(&c.CertFile, "tls-cert-file", c.CertFile, "File containing the default x509 Certificate for HTTPS.")
+	fs.StringVar(&c.KeyFile, "tls-private-key-file", c.KeyFile, "File containing the default x509 private key matching --tls-cert-file.")
+	fs.StringVar(&c.KeystoneURL, "keystone-url", c.KeystoneURL, "URL for the OpenStack Keystone API")
+	fs.StringVar(&c.KeystoneCA, "keystone-ca-file", c.KeystoneCA, "File containing the certificate authority for Keystone Service.")
+	fs.StringVar(&c.PolicyFile, "keystone-policy-file", c.PolicyFile, "File containing the policy, if provided, it takes precedence over the policy configmap.")
+	fs.StringVar(&c.PolicyConfigMapName, "policies-configmap-name", "", "ConfigMap in kube-system namespace containing the policy configuration, the ConfigMap must contain the key 'policies'")
```
Comment (zioproto) [inline pkg/identity/keystone/config.go]: policies or policy ?

File: cmd/k8s-keystone-auth/main.go (hunk 1 above)
Comment (lingxiankong) [reply cmd/k8s-keystone-auth/main.go]: policy, fixed

**Comment by dims** [comment]
/approve

some nits left?

**Comment by Fedosin** [comment]
/lgtm

There is a typo, but I can fix it in my patch. Otherwise everything looks good!

**Pull request merged**